core/kernel/tee_ta_manager.c

```diff
@@ -823,6 +842,188 @@ TEE_Result tee_ta_invoke_command(TEE_ErrorOrigin *err,
 	return res;
 }
 
+#if defined(CFG_TA_STATS)
+static TEE_Result dump_ta_memstats(struct tee_ta_session *s,
+				   uint32_t cancel_req_to,
+				   struct tee_ta_param *param)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct tee_ta_ctx *ctx = NULL;
+	struct ts_ctx *ts_ctx = NULL;
+
+	ts_ctx = s->ts_sess.ctx;
+	if (!ts_ctx)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	ctx = ts_to_ta_ctx(ts_ctx);
+
+	if (ctx->panicked)
+		return TEE_ERROR_TARGET_DEAD;
+
+	if (tee_ta_try_set_busy(ctx)) {
+		s->param = param;
+		set_invoke_timeout(s, cancel_req_to);
+		res = ts_ctx->ops->dump_mem_stats(&s->ts_sess);
+		s->param = NULL;
+		tee_ta_clear_busy(ctx);
+	} else {
+		/* Deadlock avoided */
+		res = TEE_ERROR_BUSY;
+	}
+
+	return res;
+}
+
+static void init_dump_ctx(struct tee_ta_dump_ctx *dump_ctx)
+{
+	struct tee_ta_session *sess = NULL;
+	struct tee_ta_session_head *open_sessions = NULL;
+	struct tee_ta_ctx *ctx = NULL;
+	unsigned int n = 0;
+
+	nsec_sessions_list_head(&open_sessions);
+	/*
+	 * Scan all sessions opened from secure side by searching through
+	 * all available TA instances and for each context, scan all opened
+	 * sessions.
+	 */
+	TAILQ_FOREACH(ctx, &tee_ctxes, link) {
+		unsigned int cnt = 0;
+
+		if (!is_user_ta_ctx(&ctx->ts_ctx))
+			continue;
+
+		memcpy(&dump_ctx[n].uuid, &ctx->ts_ctx.uuid,
+		       sizeof(ctx->ts_ctx.uuid));
+		dump_ctx[n].panicked = ctx->panicked;
+		dump_ctx[n].is_user_ta = is_user_ta_ctx(&ctx->ts_ctx);
+		TAILQ_FOREACH(sess, open_sessions, link) {
+			if (sess->ts_sess.ctx == &ctx->ts_ctx) {
+				if (cnt == MAX_DUMP_SESS_NUM)
+					break;
+
+				dump_ctx[n].sess_id[cnt] = sess->id;
+				cnt++;
+			}
+		}
+
+		dump_ctx[n].sess_num = cnt;
+		n++;
+	}
+}
+
+static TEE_Result dump_ta_stats(struct tee_ta_dump_ctx *dump_ctx,
+				struct tee_ta_dump_stats *dump_stats,
+				size_t ta_count)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct tee_ta_session *sess = NULL;
+	struct tee_ta_session_head *open_sessions = NULL;
+	struct tee_ta_param param = { };
+	unsigned int i = 0;
+	unsigned int j = 0;
+
+	nsec_sessions_list_head(&open_sessions);
+
+	for (i = 0; i < ta_count; i++) {
+		struct tee_ta_dump_stats *stats = &dump_stats[i];
+
+		memcpy(&stats->uuid, &dump_ctx[i].uuid,
+		       sizeof(dump_ctx[i].uuid));
+		stats->panicked = dump_ctx[i].panicked;
+		stats->sess_num = dump_ctx[i].sess_num;
+
+		/* find a session from dump context */
+		sess = NULL;
+		for (j = 0; j < dump_ctx[i].sess_num; j++) {
+			uint32_t sess_id = dump_ctx[i].sess_id[j];
+
+			sess = tee_ta_get_session(sess_id,
+						  true,
+						  open_sessions);
+			if (sess)
```

Comment:
Better to test this together with `j < dump_ctx[i].sess_num` above in the condition part of the for loop.

Reply:
maybe like this:
```c
		/* Find a session from dump context */
		for (j = 0, sess = NULL; j < dump_ctx[i].sess_num && !sess;j++)
			sess = tee_ta_get_session(dump_ctx[i].sess_id[j], true,
						   open_sessions);
```

(by the way, nitpicking on consistency, prefer an uppercase leter for inline comment first word fisrt letter, here:
```patch
-		/* find a session from dump context */
+		/* Find a session from dump context */
```

Reply:
done